packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,460 @@
+  ? Options extends RestrictedMethodOptions<RestrictedMethodParameters>
+    ? MethodImplementation
+    : never
+  : never;
+
+export type PermissionFactory<
+  Permission extends GenericPermission,
+  RequestData extends Record<string, unknown>,
+> = (
+  options: PermissionOptions<Permission>,
+  requestData?: RequestData,
+) => Permission;
+
+export type PermissionValidator<Permission extends GenericPermission> = (
+  permission: GenericPermission,
+  origin?: OriginString,
+  target?: Permission['parentCapability'],
+) => void;
+
+/**
+ * A utility type for ensuring that the given permission target key conforms to
+ * our naming conventions.
+ *
+ * See the documentation for the distinction between target names and keys.
+ *
+ * @template Key - The target key string to apply the constraint to.
+ */
+type TargetKeyConstraint<Key extends string> = Key extends `${string}_*`
+  ? Key
+  : Key extends `${string}_`
+  ? never
+  : Key extends `${string}*`
+  ? never
+  : Key;
+
+/**
+ * A type extended by any function.
+ */
+type GenericFunction = (...args: any[]) => any;
+
+/**
+ * The base permission specification type.
+ *
+ * @template TargetKey - They key of the permission target that the permission
+ * corresponds to.
+ */
+export type PermissionSpecificationBase<TargetKey extends string> = {
+  /**
+   * The target resource of the permission. In other words, at the time of
+   * writing, the RPC method name.
+   */
+  targetKey: TargetKey;
+
+  /**
+   * An array of the caveat types that may be added to instances of this
+   * permission.
+   */
+  allowedCaveats?: Readonly<NonEmptyArray<string>>;
+
+  /**
+   * The implementation of the restricted method that the permission
+   * corresponds to.
+   */
+  methodImplementation: GenericFunction;
+
+  /**
+   * The factory function used to get permission objects. Permissions returned
+   * by this function are presumed to valid, and they will not be passed to the
+   * validator function associated with this specification (if any). In other
+   * words, the factory function should validate the permissions it creates.
+   *
+   * If no factory is specified, the {@link Permission} constructor will be
+   * used, and the validator function (if specified) will be called on newly
+   * constructed permissions.
+   */
+  factory?: PermissionFactory<any, Record<string, unknown>>;
+
+  /**
+   * The validator function used to validate permissions of the associated type
+   * whenever they are mutated. The only way a permission can be legally mutated
+   * is when its caveats are modified by the permission controller.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   */
+  validator?: PermissionValidator<GenericPermission>;
+};
+
+/**
+ * An individual permission specification object. Constrained such that the
+ * given target key must be valid.
+ *
+ * @template TargetKey - The key of the permission target.
```

Comment:
These template params don't match the type. Though the T param should be renamed ideally - not a fantastic name.